FHEM/10_FS10.pm

```diff
@@ -1,502 +1,556 @@
+					}
+				}
+			} else { # dimmen
+				if ($i == 1) { # 1. Nachricht
+					$newmsg .= 'PPPPPPPPPPPPPPPP'; # 16*32400=518400 Pause . 1 Repeat (original remote control)
+					if ($dimm >= 2) {
+						$newmsg .= '#R1';
+						IOWrite($hash, 'sendMsg', $newmsg);
+						Log3 $name, 5, "$ioname: FS10 set dimm $dimm, 1. sendMsg=$newmsg";
+						$newmsg = 'P61#'; # Reset newmsg fuer 2. Nachricht
+					}
+				} else { # 2. Nachricht
+					if ($dimm == 1) {
+						$newmsg .= '#R1'; # 1 Repeat
+					} else {
+						$newmsg .= 'PPPPPP#R' . $dimm; # 6*32400=194400 Pause . Repeats
+						Log3 $name, 5, "$ioname: FS10 set dimm $dimm, 2. sendMsg=$newmsg";
+					}
+				}
+			}
+			if ($i == 2) { # 2. Nachricht
+				Log3 $name, 5, "$ioname: FS10 set sendMsg=$newmsg";
+				IOWrite($hash, 'sendMsg', $newmsg);
+			}
+		}
+	}
+
+	# Set the state of a device to off if on-for-timer is called
+	if ($modules{FS10}{ldata}{$name}) {
+		CommandDelete(undef, $name . '_timer');
+		delete $modules{FS10}{ldata}{$name};
+	}
+
+	# following timers
+	if ($setstate eq 'on' && AttrVal($name, 'follow-on-for-timer', 0)) {
+		my $dur = AttrVal($name, 'follow-on-timer', 0);
+		if ($dur > 0) {
+			my $newState = 'off';
+			my $to = sprintf '%02d:%02d:%02d', $dur/3600, ($dur%3600)/60, $dur%60;
+			Log3 $name, 3, "$ioname: FS10_set $name Set_Follow +$to setstate $newState";
+			CommandDefine(undef, $name."_timer at +$to "."setstate $name $newState; trigger $name $newState");
+			$modules{FS10}{ldata}{$name} = $to;
+		}
+	}
+
+	readingsSingleUpdate($hash, 'state', $setstate, 1);
+	return $ret;
 }
 
-#############################
-sub
-FS10_Undef($$)
-{
-  my ($hash, $name) = @_;
-  delete($modules{FS10}{defptr}{$hash->{CODE}})
-     if(defined($hash->{CODE}) &&
-        defined($modules{FS10}{defptr}{$hash->{CODE}}));
-  return undef;
+sub Define {
+	# define <name> FS10 <hauscode>_<button>
+	# define FS10_Test_1 FS10 7_11
+	# define <name> FS10 <hauscode>_<button> <iodevice>
+	# define FS10_Test_1 FS10 7_11 sduino434
+	my ($hash, $def) = @_;
+	my @a = split m{\s+}xms , $def;
+	my $name = $hash->{NAME};
+	my $iodevice;
+	my $ioname;
+
+	return 'Define FS10 wrong syntax: define <name> FS10 housecode_button' if (int(@a) < 3);
+	my ($housecode, $btncode) = split /[_]/xms , $a[2], 2;
+	return 'Define FS10 wrong syntax, must be: housecode_button' if (!defined $housecode || !defined $btncode);
+	return 'Define FS10 wrong housecode format: specify a 1 digit value [1-8]' if ($housecode !~ m/^[1-8]$/xms );
+	return 'Define FS10 wrong button format: specify a 2 digit value [0-7]' if ($btncode !~ m/^[0-7]{2}$/xms ); # Ebene Low, Ebene High
+	if (int(@a) == 4) { $iodevice = $a[3] };
+
+	$hash->{DEF} = $a[2];
+	$hash->{HC} = $housecode;
+	$hash->{BTN} = $btncode;
+	$hash->{CODE} = $a[2];
+	$hash->{VersionModule} = $VERSION;
+
+	$modules{FS10}{defptr}{$hash->{DEF}} = $hash;
+	if (exists $modules{FS10}{defptr}{ioname} && !$iodevice) { $ioname = $modules{FS10}{defptr}{ioname} };
```

Comment:
Diese und die folgende Zeile konnte ich im Diff nicht finden, wie sie vorher war.

Gab es diese da?
Wird $iodevice irgendwo mit einem Wert belegt? Ich habe es nicht gefunden :(

Reply:
Die Zeilen gab es tatsächlich noch nicht. Diese dienen dazu, gleich das richtige IODev zu setzen.
$iodevice wird in Zeile 250 zugewiesen, falls das Gerät manuell angelegt wird.

Reply:
Vielleicht ist hier dann !defined besser geeignet, den $iodevice scheint ja nur einen Wert zu haben wenn @a vier Werte enthält.

Reply:
Geht auch - geändert.